src/System.IO.Compression/tests/ZipArchive/zip_ManualAndCompatibilityTests.cs

```diff
@@ -28,5 +28,40 @@ public class zip_ManualAndCompatabilityTests
         {
             ZipTest.IsZipSameAsDir(await StreamHelpers.CreateTempCopyStream(ZipTest.compat(withTrailing)), ZipTest.compat(withoutTrailing), ZipArchiveMode.Update, dontRequireExplicit, dontCheckTimes);
         }
+
+        /// <summary>
+        /// This test ensures that a zipfile created on one platform with a file containing potentially invalid characters elsewhere
+        /// will be interpreted based on the source OS path name validation rules. 
+        /// 
+        /// For example, the file "aa\bb\cc\dd" in a zip created on Unix should be one file "aa\bb\cc\dd" whereas the same file
+        /// in a zip created on Windows should be interpreted as the file "dd" underneath three subdirectories.
+        /// </summary>
+        [Theory]
+        [InlineData("backslashes_FromUnix.zip", "aa\\bb\\cc\\dd")]
+        [InlineData("backslashes_FromWindows.zip", "dd")]
+        [InlineData("WindowsInvalid_FromUnix.zip", "aa<b>d")]
+        public static async Task ZipWithInvalidFileNames_ParsedBasedOnSourceOS(string zipName, string fileName)
+        {
+            using (Stream stream = await StreamHelpers.CreateTempCopyStream(ZipTest.compat(zipName)))
+            using (ZipArchive archive = new ZipArchive(stream))
+            {
+                Assert.Equal(1, archive.Entries.Count);
+                ZipArchiveEntry entry = archive.Entries[0];
+                Assert.Equal(fileName, entry.Name);
+            }
+        }
+
+        [Theory]
+        [InlineData("WindowsInvalid_FromWindows.zip")]
+        [InlineData("NullCharFileName_FromWindows.zip")]
+        [InlineData("NullCharFileName_FromUnix.zip")]
+        public static async Task ZipWithInvalidFileName_ThrowsArgumentException(string zipName)
+        {
+            using (Stream stream = await StreamHelpers.CreateTempCopyStream(ZipTest.compat(zipName)))
+            using (ZipArchive archive = new ZipArchive(stream))
+            {
+                Assert.Throws<ArgumentException>(() => archive.Entries);
```

Comment:
I think we want it to be a simply explainable as possible, e.g. "The type gives you the data that's in the .zip exactly as it is in the .zip, and only throws an exception if it's not possible to do the requested operation."  In other words, regardless of OS, you get the path information exactly as it comes from the .zip, but it fails when you try to extract it to disk if the path is invalid for the target OS.  Is that reasonable?